[Merge] track latest valid finalised slot

tbenr opened the pull request:
## PR Description
track latest finalised slot that has been validated against execution layer.

The tracking is done in `ForkChoice` and in `OptimisticHeadValidator` (which may be unnecessary)

## Fixed Issue(s)
related to #4671

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

## Changelog

- [x] I thought about adding a changelog entry, and added one if I deemed necessary.

tbenr committed:
track latest finalized slot that has been validated against execution layer

tbenr committed:
merges with master
fixes some bugs

tbenr committed:
fixes random field in ExecutionPayload generation in BlockProposalTestUtil
implements UTs

ajsutton commented on ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java:
```diff
@@ -272,6 +272,14 @@ private BlockImportResult importBlockAndState(
+
+    if (payloadResult.getStatus() == ExecutionPayloadStatus.VALID) {
+      UInt64 latestValidFinalizedSlot = transaction.getLatestFinalized().getSlot();
+      if (latestValidFinalizedSlot.compareTo(transaction.getLatestValidFinalizedSlot()) != 0) {
```
I think this is fine as is but we shouldn't ever let the latest valid finalised slot go backwards so I'd be tempted to use:

```suggestion
      if (latestValidFinalizedSlot.isGreaterThan(transaction.getLatestValidFinalizedSlot())) {
```

mostly just for clarity but also a little bit of defensiveness to ensure it never goes backwards.

tbenr replied:
i'm always pro defensiveness, but in this case i'd like to understand better the scenario.

AFAIK finalised anchor should never go backward, right?
Assuming that it does, why do you think we should not move back the slot accordingly? In other words, which side effect we want to avoid?

ajsutton replied:
I'm thinking of a scenario where we are concurrently performing the state transition for two blocks, one of which updates the finalised checkpoint (A) and the other doesn't (B).  Since we've done the state transition prior to jumping to the fork choice thread it's possible that block B completes validation against the old finalised checkpoint but we actually add block A to the store first, updating the checkpoint.  B will then be imported - it's no longer relevant and will be ignored by fork choice and eventually pruned but we need to make sure it doesn't lower the latest valid finalised slot when it imports.

ajsutton commented on ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java:
```diff
@@ -293,6 +301,27 @@ private BlockImportResult importBlockAndState(
+                return;
+              }
+              UInt64 latestValidFinalizedSlotInStore =
+                  recentChainData.getStore().getLatestFinalized().getSlot();
```
I think there are some race conditions here.  We might get a VALID payload for block 10, but by the time we get this executed on the fork choice thread, we've imported blocks 11, 12, 13 etc which updates the finalised checkpoint to something after block 10.  We need to be passing into this method the latest finalised slot we've validated.

tbenr replied:
right!

ajsutton commented on ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java:
```diff
@@ -293,6 +301,27 @@ private BlockImportResult importBlockAndState(
+
+              if (recentChainData
+                  .getLatestValidFinalizedSlot()
+                  .map(value -> value.compareTo(latestValidFinalizedSlotInStore) != 0)
```
This should definitely only update if the new value is greater than the old one.

ajsutton commented on ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/OptimisticHeadValidator.java:
```diff
@@ -107,10 +110,13 @@ private void verifyOptimisticHeads() {
+              recentChainData
+                  .getForkChoiceStrategy()
+                  .orElseThrow()
+                  .onExecutionPayloadResult(blockRoot, result.getStatus());
```
hmm, I'm actually thinking it might be best to move all of this into `forkChoice` and do it all on the fork choice thread. `ForkChoiceStrategy` does have locks to make it thread safe but we'd make life a lot easier to reason about if protoarray changes were all made on the `ForkChoice` thread and I think this is the one place we don't respect that currently.

ajsutton commented on storage/src/main/java/tech/pegasys/teku/storage/store/Store.java:
```diff
@@ -144,6 +145,9 @@ private Store(
     states.cache(finalizedAnchor.getRoot(), finalizedAnchor);
 
+    // [Merge] Track latest finalized slot validated against EL
+    latestValidFinalizedSlot = UInt64.ZERO;
```
nit: I'd move this initialisation to the field declaration for simplicity.

ajsutton commented on storage/src/main/java/tech/pegasys/teku/storage/client/RecentChainData.java:
```diff
@@ -526,6 +526,10 @@ public boolean isJustifiedCheckpointFullyValidated() {
 
+  public Optional<UInt64> getLatestValidFinalizedSlot() {
+    return store == null ? Optional.empty() : Optional.of(store.getLatestValidFinalizedSlot());
+  }
```
I don't think we need this to return `Optional` since we know that the genesis block is always fully valid, if we don't have a `store` yet we can just return `ZERO` and keep things simple.

ajsutton reviewed (commented):
Looking good generally. We'll need a follow up PR at some point to actually store this to disk but good to do it in memory first for now since adding a new field to the DB is hard to reverse.

tbenr committed:
safer comparisons
fixes potential concurrency issue when updating from OptimisticHeadValidator
moves ForkChoiceStrategy::onExecutionPayloadResult call in forkChoice thread
removes unneeded Optional

ajsutton reviewed (approved):
LGTM.

tbenr committed:
Merge branch 'master' into track_latest_valid_finalized_slot
